[AMDGPU] Create an AMDGPUIfConverter pass

jmmartinez opened the pull request:
This is a draft to get some early feedback about this work.

Currently the code is mined with TODOs. Some thing that are missing:
- [ ] Cost model - (`shouldConvert` function)
- [ ] Factorize more code from `EarlyIfConverter` since the 3 passes share a lot in common
- [ ] Proper implementation of `getReversedVCMPXOpcode`
- [ ] Do not remove `EarlyIfConverter` from the pipeline
- [ ] Test! Test! Test!
- [ ] Handle more conditionals
- [ ] The `needsPredication` function is probably wrong

This work is related to `SWDEV-477895`.

jmmartinez committed:
[AMDGPU][AMDGPUIfConverter] create new pass (boilerplate only)

jmmartinez committed:
[AMDGPU][AMDGPUIfConverter] Add more boilerplate (copy from EarlyIfConversion)

jmmartinez committed:
[AMDGPU][AMDGPUIfConverter] First version of AMDGPUIfConversion

jmmartinez committed:
[AMDGPU][AMDGPUIfConverter] pre commit test

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -199,11 +199,6 @@ static VGPRRegisterRegAlloc fastRegAllocVGPR(
-static cl::opt<bool>
-EnableEarlyIfConversion("amdgpu-early-ifcvt", cl::Hidden,
-                        cl::desc("Run early if-conversion"),
-                        cl::init(false));
```
I'd expect most if conversion to be done on the IR level. What cases do we need more aggressive MIR conversion for?

Can this be done in 2 steps? One showing the benefit of turning this one, and then another with the replacement pass?

jmmartinez replied:
I have one main concern about doing this at IR level: How to handle instructions accessing memory.
I guess it could be done through masked load/store operations. But what if we have some other built-in?

I'm also not sure about how to implement the cost model in the LLVM-IR, but I guess it could be possible.

jmmartinez replied:
PS: For the moment this pass hasn't covered the cases the early-ifcvt pass covers. I should add it back.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUIfConverter.cpp:
```diff
@@ -0,0 +1,279 @@
+
+namespace {
+#define DEBUG_TYPE "amdgpu-if-cvt"
+const char PassName[] = "AMDGPU if conversion";
```
Don't need the vpassname variable

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUIfConverter.cpp:
```diff
@@ -0,0 +1,279 @@
+#define DEBUG_TYPE "amdgpu-if-cvt"
+const char PassName[] = "AMDGPU if conversion";
+
+class AMDGPUIfConverter : public MachineFunctionPass {
```
Might as well start with this factored to be friendly to new PM

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUIfConverter.cpp:
```diff
@@ -0,0 +1,279 @@
+INITIALIZE_PASS_BEGIN(AMDGPUIfConverter, DEBUG_TYPE, PassName, false, false)
+INITIALIZE_PASS_DEPENDENCY(MachineDominatorTreeWrapperPass)
+INITIALIZE_PASS_DEPENDENCY(MachineBranchProbabilityInfoWrapperPass)
+INITIALIZE_PASS_END(AMDGPUIfConverter, DEBUG_TYPE, PassName, false, false)
```
Missing line at end of file

arsenm commented on llvm/test/CodeGen/AMDGPU/amdgpu-if-cvt.ll:
```diff
@@ -0,0 +1,65 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=amdgcn -mcpu=gfx1030 -verify-machineinstrs < %s | FileCheck -check-prefix=GCN %s
```
Don't need -verify-machineinstrs

arsenm commented on llvm/test/CodeGen/AMDGPU/amdgpu-if-cvt.ll:
```diff
@@ -0,0 +1,65 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=amdgcn -mcpu=gfx1030 -verify-machineinstrs < %s | FileCheck -check-prefix=GCN %s
+
+define amdgpu_kernel void @scalar_cmp(i32 noundef %value, ptr addrspace(8) nocapture writeonly %res, i32 noundef %v_offset, i32 noundef %0, i32 noundef %flag) {
```
Prefer regular functions with inreg arguments for SGPRS over kernels. Also precommit tests

nvjle commented:
I realize this is a draft patch, but this refactoring of the early if-conversion code is a great idea!

As this matures, does it makes sense to separate the target-independent portion into an isolated patch? Or at least provide a more generic title so other downstream users don't overlook the target-independent changes.

FWIW, I'm using it downstream in an NVIDIA backend-- specifically, the `EarlyIfPredicator`--  and I know of at least one other downstream backend using it. I'd be happy to review the target-independent code and try it locally on our tree. I believe the `EarlyIfPredicator` is not currently used in-tree , but the `EarlyIfConverter` is.

Thanks again for spearheading this work.

jmmartinez commented:
> As this matures, does it makes sense to separate the target-independent portion into an isolated patch? Or at least provide a more generic title so other downstream users don't overlook the target-independent changes.

Yes, I will do that. I wanted to get some quick validation of this work before being dug in.

I think I'm going to start by:
1. moving `EarlyIfConverter` and `EarlyIfPredicator` to the new pass manager.
2. factorizing the boolean argument into the `PredicationStrategy`
3. move the `shouldIfConvert` into the `PredicationStrategy`
4. move the post-dominator tree traversal and the `tryIfConvert` function into `SSAIfConv`.

Only then I can start with the AMDGPU target specific stuff.

Thanks for the feedback !

jmmartinez committed:
[NFC][EarlyIfConverter] Replace boolean Preadicate for a class

This will help when adding new EarlyIfConverter transformations (other
than the current 2)

jmmartinez committed:
[NFC][EarlyIfConverter] Make SSAIfConv into a local variable

This will simplify refactoring later.

jmmartinez committed:
[NFC][EarlyIfConverter] Move shouldConverIf in the strategy

jmmartinez committed:
[NFC][EarlyIfConverter] move tryConvertIf into SSAIfConv

jmmartinez committed:
[NFC][EarlyIfConverter] Remove unused member variables

jmmartinez committed:
[NFC][EarlyIfConverter] rearrange annonymous namespaces

jmmartinez committed:
[EarlyIfConverter] Move SSAIfConv into a spearate header/source

jmmartinez commented:
@jayfoad @arsenm 

## If Conversion update

I'm having second thoughts on the implementation of this optimization as a stand-alone optimization, instead of spreading the transformation across other passes.

### Problems 

As it is now, this pass doesn't compose nicely with `SIOptimizeExecMasking`:
To promote `vcmp+s_and_saveexec` into `vcmpx`, there should be no other instruction modifying `exec` after the comparison.
This collides with the restauration of `exec` that we have to do at the end of the `Then` block.
We can work around this by generating directly the right `vcmpx` in this transformation directly.


When I try to generalize this pass from `s_cbranch_scc0/1` to also support `s_cbranch_execz/nz`:
  * Depending on where I put the pass in the pipeline, `exec` branches may appear as `SI_IF`. To overcome this, we have to do this transformation after `SILowerControlFlow`.
  * The first terminator of the basic-block may not be the `s_cbranch_execz/nz`, but the `s_mov_b32_term` before the `s_cbranch`.
    This doesn't play well with the `SSAIfConv` pass that drops the old terminators when it moves the instructions from the `Then` block into the `Head`.
    For example:
    ```asm
    $exec_lo = S_MOV_B32_term killed %32:sreg_32 <- the terminator starts here
    S_CBRANCH_EXECZ %bb.2, implicit $exec
    S_BRANCH %bb.1
    ```

### Alternative solution

I'm thinking that it would be better to spread this transformation in 3 parts:

1. Transform `s_cbranch_scc0/1` into `vcmp + s_and_saveexec + s_cbranch_execz`
2. Let `SIOptimizeExecMasking` optimize `vcmp + s_and_saveexec` pairs into `vcmpx` if applicable
3. If-Convert `s_cbranch_execz`

#### 1. Transform `s_cbranch_scc0/1` example

Transform this:
```asm
bb.0.entry:
  ...
  S_CMP_LT_I32 killed %14:sgpr_32, 1, implicit-def $scc
  S_CBRANCH_SCC1 %bb.2, implicit killed $scc
  S_BRANCH %bb.1

bb.1.if.then:
  ...

bb.2.if.end:
  ...
```

Into this:
```asm
bb.0.entry:
  ...
  $vcc_lo = V_CMP_LT_I32_e64 killed %14:sgpr_32, 1, implicit $exec
  28:sreg_32 = S_AND_SAVEEXEC_B32 killed $vcc_lo, implicit-def $exec, implicit-def dead $scc, implicit $exec # backup exec and mask it
  S_CBRANCH_EXECZ %bb.2, implicit killed $scc
  S_BRANCH %bb.1

bb.1.if.then:
  ...

bb.2.if.end:
  $exec_lo = S_MOV_B32 killed %28:sreg_32 # restore exec
  ...
```

For `s_cbranch_scc0` we would have to reverse the comparison.

#### 3. If-Convert `s_cbranch_execz` example

Transform this:
```asm
bb.0.entry:
  ...
  S_CBRANCH_EXECZ %bb.2, implicit killed $scc
  S_BRANCH %bb.1

bb.1.if.then:
  ..Then..

bb.2.if.end:
  ..End..
```

Into this:
```asm
bb.0.entry:
  ...
  ..Then..
  ..End..
```

Keep in mind that the instructions in the `Then` block should be either speculatable or masked by `exec`, and should not modify `exec`.

## Where to go from here?

I'd appreciate some feedback if you think this is a better/worse idea; where to place this transformations (e.g. some existing pass or where in the pipeline); other potential issues that I've might have missed.

jayfoad commented:
What are you trying to achieve? Do you have a motivating example? The patch description only says what is missing from the patch.

jmmartinez commented:
@jayfoad An extract of the motivating example is this code: https://godbolt.org/z/dPdf7r381 .

This extract comes from SWDEV-477895, where CK used inline-assembly to work around code that doesn't optimize to a `v_cmpx` instruction.

jayfoad commented:
> @jayfoad An extract of the motivating example is this code: https://godbolt.org/z/dPdf7r381 .
```
        s_cmp_lt_i32 s0, 1
        s_cbranch_scc1 .LBB0_2
        s_clause 0x1
        s_load_dwordx2 s[6:7], s[4:5], 0x0
        s_load_dwordx4 s[0:3], s[4:5], 0x10
        s_waitcnt lgkmcnt(0)
        s_load_dword s6, s[6:7], 0x0
        s_load_dword s4, s[4:5], 0x20
        s_waitcnt lgkmcnt(0)
        v_mov_b32_e32 v0, s6
        v_mov_b32_e32 v1, s4
        buffer_store_dword v0, v1, s[0:3], 0 offen
.LBB0_2:
```
So you want to remove the `s_cbranch_scc1` and instead rely on setting EXEC to 0, even though the branch condition is uniform? Is that really viable, given that the s_load instructions can't be predicated with EXEC? What codegen would you like to see for this example?

jmmartinez commented:
Something like this:

```asm
        v_mov_b32_e32 exec_backup, exec
        v_cmpx_lt_i32 exec, s0, 1
        s_load_dwordx2 s[6:7], s[4:5], 0x0
        s_load_dwordx4 s[0:3], s[4:5], 0x10
        s_waitcnt lgkmcnt(0)
        s_load_dword s6, s[6:7], 0x0
        s_load_dword s4, s[4:5], 0x20
        s_waitcnt lgkmcnt(0)
        v_mov_b32_e32 v0, s6
        v_mov_b32_e32 v1, s4
        buffer_store_dword v0, v1, s[0:3], 0 offen
        v_mov_b32_e32 exec, exec_backup
```

> Is that really viable, given that the s_load instructions can't be predicated with EXEC?

This depends if the loaded pointers are marked as dereferenceable. If they are we can safely execute them.

jmmartinez commented:
Thanks ! I didn't know about `SIPreEmitPeephole::removeExeczBranch`. I'll try to approach this problem from a different angle.
